geode-wan/src/test/java/org/apache/geode/cache/client/internal/locator/wan/LocatorMembershipListenerTest.java

```diff
@@ -0,0 +1,313 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.geode.distributed.internal.DistributionConfig;
+import org.apache.geode.distributed.internal.tcpserver.TcpClient;
+import org.apache.geode.internal.admin.remote.DistributionLocatorId;
+
+public class LocatorMembershipListenerTest {
+  private TcpClient tcpClient;
+  private LocatorMembershipListenerImpl locatorMembershipListener;
+
+  private DistributionLocatorId buildDistributionLocatorId(int port) {
+    return new DistributionLocatorId("localhost[" + port + "]");
+  }
+
+  private List<LocatorJoinMessage> buildPermutationsForClusterId(int dsId, int locatorsAmount) {
+    int basePort = dsId * 10000;
+    List<LocatorJoinMessage> joinMessages = new ArrayList<>();
+
+    for (int i = 1; i <= locatorsAmount; i++) {
+      DistributionLocatorId sourceLocatorId = buildDistributionLocatorId(basePort + i);
+
+      for (int j = 1; j <= locatorsAmount; j++) {
+        DistributionLocatorId distributionLocatorId = buildDistributionLocatorId(basePort + j);
+        LocatorJoinMessage locatorJoinMessage =
+            new LocatorJoinMessage(dsId, distributionLocatorId, sourceLocatorId, "");
+        joinMessages.add(locatorJoinMessage);
+      }
+    }
+
+    return joinMessages;
+  }
+
+  private void joinLocatorsDistributorThread() throws InterruptedException {
+    Set<Thread> threads = Thread.getAllStackTraces().keySet();
+    Optional<Thread> distributorThread = threads.stream()
+        .filter(t -> WAN_LOCATORS_DISTRIBUTOR_THREAD_NAME.equals(t.getName()))
+        .findFirst();
+
+    if (distributorThread.isPresent()) {
+      distributorThread.get().join();
+    }
+  }
+
+  @Before
+  public void setUp() {
+    DistributionConfig distributionConfig = mock(DistributionConfig.class);
+    when(distributionConfig.getStartLocator()).thenReturn(DistributionConfig.DEFAULT_START_LOCATOR);
+    when(distributionConfig.getMemberTimeout())
+        .thenReturn(DistributionConfig.DEFAULT_MEMBER_TIMEOUT);
+
+    tcpClient = spy(mock(TcpClient.class));
```

Comment:
There is no need to wrap a `mock` inside a `spy`. In fact, `Mockito` mocks provide the same abilities to verify invocations as a spy does.

There's never a need to use both `mock` and `spy` on the same field or variable:
```
TcpClient mockTcpClient = mock(TcpClient.class);

TcpClient realTcpClient1 = spy(TcpClient.class); // invokes `new TcpClient()`
TcpClient realTcpClient2 = spy(new TcpClient());
TcpClient realTcpClient3 = spy(new TcpClient(mock(DistributionConfig.class)));
TcpClient realTcpClient4 = spy(new TcpClient(new Properties()));
TcpClient realTcpClient5 = spy(new TcpClient(mock(SocketCreator.class)));
```

Reply:
Sounds good, thank you!.